### PR TITLE
mgr/cephadm: allow monitoring stack ip binding and custom ports

### PR DESCRIPTION
**Pull request by Daniel-Pivonka**
Fixes: https://tracker.ceph.com/issues/50359
Signed-off-by: Daniel Pivonka <dpivonka@redhat.com>

### CODE REVIEW

**Comment by Daniel-Pivonka** [comment]
works but making this a draft for now while i test more and clean up code

File: src/python-common/ceph/deployment/service_spec.py [hunk 1]
```diff
@@ -1004,3 +1025,37 @@ def config_json(self) -> Dict[str, Any]:
...
+            networks=networks)
+
+        self.service_type = service_type
+        self.port = port
+
+    def get_port_start(self) -> List[int]:
+        return [self.get_port()]
+
+    def get_port(self) -> int:
+        if self.port:
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/service_spec.py:1056]:
```suggestion
    @property
    def port(self) -> int:
        if self._port:
```
?

**Review by sebastian-philipp — APPROVED** [review]

**Review by adk3798 — APPROVED** [review]

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]
can you add a test for this?

File: src/python-common/ceph/deployment/service_spec.py (hunk 1 above)
Comment (Daniel-Pivonka) [reply src/python-common/ceph/deployment/service_spec.py:1056]: all other services classes have a function called "get_port". leaving this how it is to match the other services

**Review by sebastian-philipp — APPROVED** [review]

File: doc/cephadm/monitoring.rst [hunk 2]
```diff
@@ -128,6 +128,24 @@ update its configuration:
...
+.. code-block:: yaml
+
+    service_type: grafana
+    service_name: grafana
+    placement:
+      count: 1
+    networks:
+    - 192.169.142.0/24
+    spec:
+      port: 4200
```
Comment (p-se) [inline doc/cephadm/monitoring.rst:147]: Is it also possible to specify the port in the CLI command `ceph orch apply grafana ...`?

**Review by p-se — CHANGES_REQUESTED** [review]
When changing the port for the node-exporter and then deploying Prometheus, Prometheus' configuration file will point to port 9100 for the node-exporter and not for what is actually listening on. 

Probably [this line](https://github.com/Daniel-Pivonka/ceph/blob/cephadm-monitoring-ip-bind/src/pybind/mgr/cephadm/services/monitoring.py#L241) need to adapted, too.

File: doc/cephadm/monitoring.rst [hunk 3]
```diff
@@ -128,6 +128,24 @@ update its configuration:
 
 The ``reconfig`` command also sets the proper URL for Ceph Dashboard.
 
+Networks and Ports
+~~~~~~~~~~~~~~~~~~
+
+All monitoring services can have the network and port they bind to configured with a yaml service specification
```
Comment (p-se) [inline doc/cephadm/monitoring.rst:134]: Does that include the alertmanager?

Comment (p-se) [reply doc/cephadm/monitoring.rst:134]: nvm, just seen that the AlertManagerSpec already exists!

**Comment by Daniel-Pivonka** [comment]
all configs get populated with the correct ip:port when using custom network/port for monitoring services

File: doc/cephadm/monitoring.rst (hunk 2 above)
Comment (Daniel-Pivonka) [reply doc/cephadm/monitoring.rst:147]: not at the moment. that would require adding a '--port' option to the generic 'ceph orch apply' command but that option would not be useful for all the services that can be applied with 'ceph orch apply'

**Review by p-se — APPROVED** [review]
lgtm

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-06-05_02:55:59-rados-wip-kefu-testing-2021-06-05-0940-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-08_12:29:52-rados-wip-kefu-testing-2021-06-08-1034-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/50151
- https://tracker.ceph.com/issues/51074

**Comment by fmount** [comment]
@sebastian-philipp  Is this patch going to be backported in -pacific?

**Comment by sebastian-philipp** [comment]
of course

**Pull request merged**